File: apps/web/src/lib/localized.ts
// Зеркалит GameOrg.Api.Common.LocalizedTextDto — {az,ru,en} вместо голого
// Dictionary, иначе Kiota на нём сворачивает всё в additionalData-мешок
// (так сейчас выглядит nameI18n у городов).
export type LocalizedText = {
  az?: string | null;
  ru?: string | null;
  en?: string | null;
};

export const EMPTY_LOCALIZED_TEXT: LocalizedText = { az: "", ru: "", en: "" };

const FALLBACK_ORDER: Array<keyof LocalizedText> = ["az", "en", "ru"];

/**
 * Тот же фолбэк, что GameOrg.Api.Common.Localized.Resolve на бэкенде:
 * запрошенный язык → az → en → ru → любой непустой. На практике почти не
 * нужен — резолвнутую строку почти всегда уже прислал бэкенд, но держим для
 * мест, где на руках только сырой LocalizedText (например, сразу после
 * сохранения формы, до похода на сервер).
 */
export function pickLocalized(value: LocalizedText | null | undefined, locale: string): string {
  if (!value) return "";

  const direct = value[locale as keyof LocalizedText];
  if (direct) return direct;

  for (const lang of FALLBACK_ORDER) {
    const v = value[lang];
    if (v) return v;
  }

  for (const v of Object.values(value)) {
    if (v) return v;
  }

  return "";
}

File: apps/api/src/GameOrg.Api/Common/Localized.cs
namespace GameOrg.Api.Common;

/// <summary>
/// Фолбэк для мультиязычных полей (Venue.NameI18n, User.DisplayNameI18n, ...):
/// запрошенный язык → az → en → ru → первый непустой. Решение пользователя —
/// никогда не показывать пустоту вместо текста, и не помечать, что это
/// текст на другом языке (см. docs/PLAN.md, Шаг 7.5).
/// </summary>
public static class Localized
{
    // az первым — дефолт сайта; дальше en (не ru) — так решил пользователь.
    private static readonly string[] FallbackOrder = ["az", "en", "ru"];

    /// <summary>Пустые строки считаются отсутствующими — форма могла сохранить "" вместо null.</summary>
    public static string? Resolve(IReadOnlyDictionary<string, string>? values, string locale)
    {
        if (values is null) return null;

        if (TryGetNonEmpty(values, locale, out var direct)) return direct;

        foreach (var fallback in FallbackOrder)
            if (TryGetNonEmpty(values, fallback, out var value))
                return value;

        foreach (var value in values.Values)
            if (!string.IsNullOrEmpty(value))
                return value;

        return null;
    }

    private static bool TryGetNonEmpty(IReadOnlyDictionary<string, string> values, string key, out string value)
    {
        value = values.TryGetValue(key, out var v) ? v : "";
        return !string.IsNullOrEmpty(value);
    }
}
